src/gui/programupdater.cpp

```diff
@@ -134,51 +159,11 @@ void ProgramUpdater::rssDownloadFinished(const Net::DownloadResult &result)
         }
     }
 
-    emit updateCheckFinished(!m_updateUrl.isEmpty(), version, m_invokedByUser);
-}
-
-void ProgramUpdater::updateProgram()
-{
-    Q_ASSERT(!m_updateUrl.isEmpty());
-    QDesktopServices::openUrl(m_updateUrl);
-    return;
+    emit updateCheckFinished(!m_updateURL.isEmpty(), version);
 }
 
-bool ProgramUpdater::isVersionMoreRecent(const QString &remoteVersion) const
+void ProgramUpdater::updateProgram() const
 {
-    const QRegularExpressionMatch regVerMatch = QRegularExpression("([0-9.]+)").match(QBT_VERSION);
-    if (regVerMatch.hasMatch())
-    {
-        const QString localVersion = regVerMatch.captured(1);
-        const QVector<QStringRef> remoteParts = remoteVersion.splitRef('.');
-        const QVector<QStringRef> localParts = localVersion.splitRef('.');
-
-        for (int i = 0; i < qMin(remoteParts.size(), localParts.size()); ++i)
-        {
-            if (remoteParts[i].toInt() > localParts[i].toInt())
-                return true;
-            if (remoteParts[i].toInt() < localParts[i].toInt())
-                return false;
-        }
-        // Compared parts were equal, if remote version is longer, then it's more recent (2.9.2.1 > 2.9.2)
-        if (remoteParts.size() > localParts.size())
-            return true;
-        // versions are equal, check if the local version is a development release, in which case it is older (2.9.2beta < 2.9.2)
-        const QRegularExpressionMatch regDevelMatch = QRegularExpression("(alpha|beta|rc)").match(QBT_VERSION);
-        if (regDevelMatch.hasMatch())
-            return true;
-    }
-    return false;
-}
-
-namespace
-{
-    QString getStringValue(QXmlStreamReader &xml)
-    {
-        xml.readNext();
-        if (xml.isCharacters() && !xml.isWhitespace())
-            return xml.text().toString();
-
-        return {};
-    }
+    Q_ASSERT(!m_updateURL.isEmpty());
```

Comment:
Corresponding checking method (aka `isUpdateAvailable()`) could make this aspect of ProgramUpdater interface complete.

Reply:
>Corresponding checking method (aka isUpdateAvailable()) could make this aspect of ProgramUpdater interface complete.

IMO there are 2 ways to provide a clean interface. One is you suggested, putting all query functions to ProgramUpdater public interface. Another one is the first edition of this PR, using std::function to wrap all possible follow-up actions and pass it to caller in the corresponding signal and ProgramUpdater would only need to provide one public function: check if any new version exists.
I settled on the first one since the second way didn't receive positive feedbacks.